### ISSUE TITLE
On TB 115 messages are only sent from Local Folders

### ISSUE THREAD

**Issue by quaqo** [Extended-Thunder/send-later#652]
### Prerequisites

- [X] I am reporting a problem, not asking a question (for questions, go to [Discussions](https://github.com/Extended-Thunder/send-later/discussions)).
- [X] I have consulted the [user guide](https://extended-thunder.github.io/send-later/).
- [X] I have confirmed that I'm running the current version and the problem persists there (note the version number in fine print next to the add-on name [on addons.thunderbird.net](https://addons.thunderbird.net/thunderbird/addon/send-later-3/) and compare it to the version number listed in the add-on details opened from the Add-ons page in Thunderbird; if you've got an old version, click the gear icon and select "Check for Updates").
- [X] If I'm having an issue with date-parsing confusion, I've read the [section about that in the user guide](https://extended-thunder.github.io/send-later/#date-format-confusion-in-the-scheduler-pop-up) and I still think this is something different.
- [X] If I'm having an issue with messages not being delivered, I've followed the [troubleshooting steps in the user guide](https://extended-thunder.github.io/send-later/#messages-dont-send-or-send-multiple-times) and it's still an issue.

### Describe the bug

This week I've been upgraded to TB 115 (115.3.1 on Win64). Since then SL (10.2.10) doesn't work properly with IMAP accounts.

Messages are being placed in the account's draft folder, but they are then completely hidden to the Extension.

I understand that on 115 there's no panel view and header view is it or miss, but when clicking on the button I should still get a list: this doesn't happen. Looking at the console output there are 0 scheduled messages also.

But if I manually move a SL message from a IMAP draft folder to the Local Folders/Drafts folder, SL becomes aware of it! And the message then gets properly sent out at the right time.

Also, if I activate the subfolder option, the subfolder is correctly created on the IMAP server, but the message it is still placed in the parent Draft folder.

### Operating System

Windows

### Operating System Version

10

### Thunderbird version

115.3.1

**Comment by jikamens** [comment]
Are you by any chance using the "Manually Sort Folders" add-on, or have you used it in the past?
Can you try disabling your other add-ons and seeing if that makes the problems go away?

**Comment by quaqo** [comment]
I never had that add-on. I disabled everything but SL. I confirm that full functionality is achieved only in Local Folders. I have 2 Gmail accounts and 1 Outlook365 account, plus another general IMAP account and it doesn't work in their Draft folders.

If I move the very same message in Local Folders/Drafts it gets picked up (I can see the header, I can see it in the SL list by clicking on the button, etc).

An interesting bit in the log could be:

```
Found Draft folder(s) for account Gmail1-IMAP       Array []

Found Draft folder(s) for account Gmail2-IMAP       Array []
Found Draft folder(s) for account Outlook365-IMAP       Array []
Found Draft folder(s) for account My-IMAP      Array []
Found Draft folder(s) for account Local Folders 
      Array [ {…} ]
  0: Object { accountId: "account2", name: "Drafts", path: "/Drafts", … }
```

**Comment by jikamens** [comment]
That is indeed very interesting. From that log snippet Send Later simply isn't seeing the Drafts folders associated with your Gmail accounts.

We are relying on Thunderbird to tell us which folders are Drafts folders:

https://github.com/Extended-Thunder/send-later/blob/main/utils/tools.js#L189

Thunderbird does not appear to be returning correct information to Send Later.

How do you have your Drafts folders configured in the Account Settings for these accounts?

**Comment by quaqo** [comment]
> How do you have your Drafts folders configured in the Account Settings for these accounts?

I have them specifically selected as "Drafts" in Account Settings. I don't let TB autoselect them. It's been like this since forever and always worked well with SL.

So you do confirm that could be the problem. If so, I'm going to investigate a bit more and report back.

**Comment by JonathanKamensDVA** [comment]
Well, yes, that could be what's causing the issue, but even if so, I'd argue TB is still behaving incorrectly so it should still be reported to the TB team as a bug. But let's see what your investigation reveals, since that will help us give the TB team a more thorough bug report.

**Comment by jikamens** [comment]
Well, yes, that could be what's causing the issue, but even if so, I'd argue TB is still behaving incorrectly so it should still be reported to the TB team as a bug. But let's see what your investigation reveals, since that will help us give the TB team a more thorough bug report.

**Comment by quaqo** [comment]
> Well, yes, that could be what's causing the issue, but even if so, I'd argue TB is still behaving incorrectly so it should still be reported to the TB team as a bug. But let's see what your investigation reveals, since that will help us give the TB team a more thorough bug report.

Ok so I found the issue. I have them selected also as "templates" and it seems that this causes 115 to only report them as such in WebExtension. This is not the case in TB UI (I can group view them both as "Drafts" and "Templates").

After I selected another folder as a Template sometimes it was still reporting them as "templates" (randomly).

I traced this back to having multiple identities for each account and I then did the same for each identity.

After this TB is reporting Drafts as type "draft" again (like it did in previous versions).

One thing that's not working is the SubFolder option. I don't know if it is a Gmail problem. I realised that SL is iterating and timouting alot on async calls because I have 200+ draft messages in Gmail. It's very slow compared to before. A working subfolder option where it just iterates thru that limited set of messages would indeed be handy.

**Comment by quaqo** [comment]
> Ok so I found the issue. I have them selected also as "templates" and it seems that this causes 115 to only report them as such in WebExtension. This is not the case in TB UI (I can group view them both as "Drafts" and "Templates").

https://github.com/Extended-Thunder/send-later/issues/512 could be related.

>  I realised that SL is iterating and timouting alot on async calls because I have 200+ draft messages in Gmail. It's very slow compared to before.

Mayne because of https://github.com/Extended-Thunder/send-later/commit/5c0753873abab33fc6b5c42707d82ba115421012 ?

**Comment by jikamens** [comment]
@quaqo do you have either your entire Gmail account or at least the Drafts folder configured in Thunderbird to sync locally? If not, I believe that setting that up will greatly improve Send Later's speed and reduce the network and CPU load it puts on your computer. If you do have Drafts set up to sync locally and you're still seeing Send Later take a long time to iterate over all your drafts, then I definitely want to hear about that, because in that case I really need to figure out how to do something about it.
Also, what do you mean exactly by "timouting alot"? Can you elaborate on that?
Regarding having the same folder configured as both drafts and templates, I don't think that's a TB change, I think that's a change in Send Later. I need to think about how I'm going to fix it, so this ticket should stay open to ensure that I don't forget to do that. If TB supports that use case then Send Later needs to as well.

**Comment by jikamens** [comment]
Just to be clear: the subfolder functionality isn't working for you even after you separate templates and drafts into separate folders?

**Comment by quaqo** [comment]
>  do you have either your entire Gmail account or at least the Drafts folder configured in Thunderbird to sync locally? If not, I believe that setting that up will greatly improve Send Later's speed and reduce the network and CPU load it puts on your computer.

I did't. I do now.

> If you do have Drafts set up to sync locally and you're still seeing Send Later take a long time to iterate over all your drafts, then I definitely want to hear about that, because in that case I really need to figure out how to do something about it. Also, what do you mean exactly by "timouting alot"? Can you elaborate on that?

I'm seeing an async fail with some "Error processing message" in the log.

> Regarding having the same folder configured as both drafts and templates, I don't think that's a TB change, I think that's a change in Send Later. I need to think about how I'm going to fix it, so this ticket should stay open to ensure that I don't forget to do that. If TB supports that use case then Send Later needs to as well.

Are you sure of that? In the debug inspecting the folder object only "template" was given, so SL was ignoring them.

**Comment by quaqo** [comment]
> Just to be clear: the subfolder functionality isn't working for you even after you separate templates and drafts into separate folders?

Correct. I'll experiment with this again either tomorrow evening (CET) or on Sunday and report back.

**Comment by quaqo** [comment]
> I did't. I do now.

It seems already faster. I'll report back on the async fails as was probabily related to this. Didn't happen before thou. Maybe because of the 100 draft limit bug?

**Comment by jikamens** [comment]
Tracking the issue about errors processing messages in #654 instead of here.

**Comment by jikamens** [comment]
>Are you sure of that? In the debug inspecting the folder object only "template" was given, so SL was ignoring them.

Yes, I'm sure, because the logic for how draft folders are identified was changed in this release.

Previously I was going "under the hood" and doing it in "Experiment" code with full access to TB internals. In this release I switched over to using the supported API for this purpose. As we can see, that may have been an unwise change. I think this is probably going to have to be addressed in the short term by going back to the experiment code, and in the long term by changes to the API, for which I will be dependent on the TB team.

**Comment by jikamens** [comment]
Tracking the subfolder not working issue in #655 instead of here.

**Comment by jikamens** [comment]
Reported to TB team [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1857631).

**Comment by jikamens** [comment]
This issue should be fixed in release 10.3.1 of Send Later.

To check what version of Send Later you have:

* Open the Add-ons page in Thunderbird ("Add-ons and Themes" from the corner "hamburger" menu).
* Click on Extensions on the left if you're not already on the left if you're not already there.
* Find Send Later in the list of extensions and click on it to open its details page.
* The version number should be displayed there next to
"Version".

This release is not yet available for download from addons.thunderbird.net, but you can download and install it from its [GitHub release page](https://github.com/Extended-Thunder/send-later/releases/tag/v10.3.1) as follows:

* Download [send_later_10.3.1.xpi](https://github.com/Extended-Thunder/send-later/releases/download/v10.3.1/send_later_10.3.1.xpi) to your computer.
(If you are using Firefox, make sure to right click on the link and select "Save Link As..." rather than just clicking on it, because otherwise Firefox will try to install it as a Firefox add-on.)
* Open the Add-ons page in Thunderbird ("Add-ons and Themes" from the corner "hamburger" menu).
* Click the gear icon at the top and select "Install Add-on From File...".
* Browse to and select the downloaded file.
* Click through the installation pop-ups.
* You can delete the downloaded file once you've installed it into Thunderbird.

To subscribe to future beta releases (we love beta testers!), download and install [send_later_beta_10.3.1.xpi](https://github.com/Extended-Thunder/send-later/releases/download/v10.3.1/send_later_10.3.1_beta.xpi) instead. The advantage is that if there's a bug that affects your workflow you'll help find it and get it fixed quickly. The disadvantage is that bugs are a bit more likely in beta releases. You can unsubscribe from beta releases at any time by downloading and installing send_later.xpi or installing from [addons.thunderbird.net](https://addons.thunderbird.net/thunderbird/addon/send-later-3/).

Note that this is a prerelease. While we make every effort to ensure that prereleases are stable, they are a bit more likely to have bugs, so proceed with caution.

If you prefer to wait, the release will eventually be available on addons.thunderbird.net, and unless you've turned off automatic add-on updates, Thunderbird will eventually update to the new version automatically. To update immediately once it's available:

* Go to [addons.thunderbird.net](https://addons.thunderbird.net/thunderbird/addon/send-later-3/versions/) and confirm that release 10.3.1 is available for download there.
* Open the Add-ons page in Thunderbird ("Add-ons and Themes" from the corner "hamburger" menu).
* Click on Send later.
* Click the gear icon and select "Check for Updates".
* The add-on should update and then you should see the message, "Your add-ons have been updated," next to the gear icon.

Please let me know if you still see this issue in this release.

**Comment by jikamens** [comment]
I'm leaving this issue open because I would like to fix it "properly" once the add-on API is fixed so that I no longer have to use an experiment to work around the API issues.

**Comment by jikamens** [comment]
Starting in TB120 in Manifest Version 3 the API returns an array of folder types rather than a single folder type, so we should be able to eliminate the experiment.

**Comment by jikamens** [comment]
Waiting for a migration guide to be published at https://developer.thunderbird.net/add-ons/manifest-v3 before attempting to upgrade to Manifest Version 3.
